linkerd/app/core/src/svc.rs

```diff
@@ -233,6 +235,47 @@ impl<S> Stack<S> {
         }))
     }
 
+    pub fn push_discover_cache<D, K, T>(
+        self,
+        discover: D,
+        idle: Duration,
+    ) -> Stack<
+        crate::disco_cache::NewDiscoveryCache<
+            K,
+            impl NewService<
+                    K,
+                    Service = impl Service<
+                        (),
+                        Error = Error,
+                        Response = D::Response,
+                        Future = impl Send + 'static,
+                    > + Clone
+                                  + Send
+                                  + Sync
+                                  + 'static,
+                > + Clone
+                + Send
+                + Sync
+                + 'static,
+            S,
+        >,
+    >
+    where
+        T: Param<K> + Clone,
+        K: Clone + fmt::Debug + Eq + Hash + Send + Sync + 'static,
+        D: Service<K> + Clone + Send + Sync + 'static,
+        D::Error: Into<Error> + Send + 'static,
+        D::Future: Send + 'static,
+        D::Response: Clone + Send + Sync + 'static,
+        S: NewService<(D::Response, T)> + Clone,
```

Comment:
I think in general we want a double-new layer here. We can always lift a double to a single but we can't go the other way.